ghcide/src/Development/IDE/Types/Logger.hs

```diff
@@ -290,3 +303,46 @@ textWithPriorityToText columns WithPriority{ priority, callStack_, payload } = d
           pure (threadIdToText threadId)
         PriorityColumn -> pure (priorityToText priority)
         DataColumn -> pure payload
+
+-- | Given a 'Recorder' that requires an argument, and an 'MVar' that
+-- will eventually be filled with that argument, produces a 'Recorder'
+-- that queues up messages until the argument is available, at which
+-- point it sends the backlog.
+withBacklog :: MVar v -> (v -> Recorder a) -> IO (Recorder a)
+withBacklog argVar recFun = do
+  backlog <- newTQueueIO
+  pure $ Recorder $ \it -> do
+        marg <- liftIO $ tryReadMVar argVar
+        case marg of
+          Nothing -> liftIO $ atomically $ writeTQueue backlog it
+          Just arg -> do
+              let recorder = recFun arg
+              toRecord <- liftIO $ atomically $ flushTQueue backlog
+              for_ toRecord (logger_ recorder)
+              logger_ recorder it
```

Comment:
So the backlog doesn't get flushed until the recorder gets called, right?  
This isn't good, specially if the recorder only gets called on error messages

Reply:
Great point! All the more reason to have a callback: then I think we can indeed do this more nicely.